elfpy/utils/apeworx_integrations.py

```diff
@@ -185,51 +153,80 @@ def ape_close_position(
-            tx_receipt = hyperdrive_contract.closeShort(  # type: ignore
-                maturity_time,
-                trade_bond_amount,
-                min_output,
-                agent_address,
-                as_underlying,
-            )
-        elif trade_prefix == hyperdrive_assets.AssetIdPrefix.LP:
-            tx_receipt = hyperdrive_contract.removeLiquidity(  # type: ignore
-                trade_bond_amount,
-                min_output,
-                agent_address,
-                as_underlying,
-            )
-        else:
-            raise ValueError(f"{trade_prefix=} must be LP (0), long (1), or short (2)")
-        # Return the updated pool state & transaction result
-        pool_state = get_pool_state(tx_receipt, hyperdrive_contract)
-    return pool_state, tx_receipt
+
+    info = {
+        "OPEN_LONG": Info(method=hyperdrive.openLong, prefix=hyperdrive_assets.AssetIdPrefix.LONG),
+        "CLOSE_LONG": Info(method=hyperdrive.closeLong, prefix=hyperdrive_assets.AssetIdPrefix.LONG),
+        "OPEN_SHORT": Info(method=hyperdrive.openShort, prefix=hyperdrive_assets.AssetIdPrefix.SHORT),
+        "CLOSE_SHORT": Info(method=hyperdrive.closeShort, prefix=hyperdrive_assets.AssetIdPrefix.SHORT),
+        "ADD_LIQUIDITY": Info(method=hyperdrive.addLiquidity, prefix=hyperdrive_assets.AssetIdPrefix.LP),
+        "REMOVE_LIQUIDITY": Info(method=hyperdrive.removeLiquidity, prefix=hyperdrive_assets.AssetIdPrefix.LP),
+    }
+    if trade_type in {"CLOSE_LONG", "CLOSE_SHORT"}:
+        assert maturity_time, "Maturity time must be provided to close a long or short trade"
+        trade_asset_id = hyperdrive_assets.encode_asset_id(info[trade_type].prefix, maturity_time)
+        amount = np.clip(amount, 0, hyperdrive.balanceOf(trade_asset_id, agent))
+    params = {
+        "_asUnderlying": True,  # mockHyperdriveTestNet does not support as_underlying=False
+        "_destination": agent,
+        "_contribution": amount,
+        "_shares": amount,
+        "_baseAmount": amount,
+        "_bondAmount": amount,
+        "_minOutput": 0,
+        "_maxDeposit": amount,
+        "_minApr": 0,
+        "_maxApr": int(100 * 1e18),
+        "agent_contract": agent,
+        "trade_amount": amount,
+        "maturation_time": maturity_time,
+    }
+    selected_abi, args = select_abi(params=params, method=info[trade_type].method)
+    contract_txn: ContractTransaction = ContractTransaction(abi=selected_abi, address=hyperdrive.address)
+
+    try:
+        tx_receipt = attempt_txn(agent, contract_txn, *args, **kwargs)
+        return get_pool_state(tx_receipt=tx_receipt, hyperdrive_contract=hyperdrive), tx_receipt  # type: ignore
+    except Exception as exc:
+        pool = hyperdrive.getPoolInfo().__dict__
+        string = "Failed to execute trade: %s\n => %s of %s\n => Agent: %s\n => Pool: %s\n"
+        var = trade_type, fmt(amount), exc, agent, pool
+        failblog(string, *var)
+        return None, None
+
+
+def attempt_txn(
```

Comment:
this function is very hard for me to follow.  lots of comments please.

Reply:
many comments were added